File: Programmers/Lv3/immigration.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    sort(times.begin(), times.end());
    long long min_time, max_time, avg_time, people;
    
    min_time=1;
    max_time = times.back() * (long long)n;
    
    while(min_time <= max_time){
        people = 0;         
        avg_time = (min_time + max_time) / 2;
        for(auto t : times) people += ( avg_time / t );
        if(people >= n) {
            answer = avg_time; 
            max_time = avg_time - 1;
        }
        else min_time = avg_time + 1;
    }
    return answer;
}

File: java/src/programmers/lv3/Immigration.java
package programmers.lv3;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

public class Immigration {
    public long solution(int n, int[] times) {
        long answer=0,avg,min,max;
        Arrays.sort(times);
        min = 0; max =  (long) times[times.length - 1] * n;

        int cnt;
        while (min <= max) {
            avg = (min + max) / 2;
            cnt = 0;
            for(int time : times) cnt += avg/time;

            if(cnt >= n) {
                answer =avg;
                max = avg - 1;
            }
            else min = avg + 1;
        }
        return answer;
    }
    @Test
    void test(){
        Assertions.assertEquals(28, solution(6,new int[]{7,10}));
    }
}
